crates/sol-types/src/types/encodable.rs

```diff
@@ -0,0 +1,405 @@
+            type SolType = $sol;
+        }
+    )*};
+}
+
+impl_encodable! {
+    // Basic
+    [] bool => sol_data::Bool [];
+
+    [] i8 => sol_data::Int::<8> [];
+    [] i16 => sol_data::Int::<16> [];
+    [] i32 => sol_data::Int::<32> [];
+    [] i64 => sol_data::Int::<64> [];
+    [] i128 => sol_data::Int::<128> [];
+    [] I256 => sol_data::Int::<256> [];
+    #[cfg(pointer_width = "32")]
+    [] isize => sol_data::Int::<32> [];
+    #[cfg(pointer_width = "64")]
+    [] isize => sol_data::Int::<64> [];
+
+    // TODO: Array<u8> is specialized to encode as `bytes[N]`
+    // [] u8 => sol_data::Uint::<8> [];
+    [] u16 => sol_data::Uint::<16> [];
+    [] u32 => sol_data::Uint::<32> [];
+    [] u64 => sol_data::Uint::<64> [];
+    [] u128 => sol_data::Uint::<128> [];
+    [] U256 => sol_data::Uint::<256> [];
+    #[cfg(pointer_width = "32")]
+    [] usize => sol_data::Uint::<32> [];
+    #[cfg(pointer_width = "64")]
+    [] usize => sol_data::Uint::<64> [];
+
+    [] Address => sol_data::Address [];
+    [] Function => sol_data::Function [];
+    [const N: usize] FixedBytes<N> => sol_data::FixedBytes<N> [where ByteCount<N>: SupportedFixedBytes];
+    [const N: usize] [u8; N] => sol_data::FixedBytes<N> [where ByteCount<N>: SupportedFixedBytes];
+    [] String => sol_data::String [];
+    [] str => sol_data::String [];
+    [] Bytes => sol_data::Bytes [];
+
+    [] Vec<u8> => sol_data::Bytes [];
+    [] [u8] => sol_data::Bytes [];
+
+    // Generic
+    [T: Encodable] Vec<T> => sol_data::Array<T::SolType> [];
+    [T: Encodable] [T] => sol_data::Array<T::SolType> [];
+    [T: Encodable, const N: usize] [T; N] => sol_data::FixedArray<T::SolType, N> [];
+
+    [T: ?Sized + Encodable + SolTypeEncodable<T::SolType>] &T => T::SolType [];
+    [T: ?Sized + Encodable + SolTypeEncodable<T::SolType>] &mut T => T::SolType [];
+    [T: ?Sized + Encodable + SolTypeEncodable<T::SolType>] alloc::boxed::Box<T> => T::SolType [];
+    [T: ?Sized + alloc::borrow::ToOwned + Encodable + SolTypeEncodable<T::SolType>] alloc::borrow::Cow<'_, T> => T::SolType [];
+    [T: ?Sized + Encodable + SolTypeEncodable<T::SolType>] alloc::rc::Rc<T> => T::SolType [];
+    [T: ?Sized + Encodable + SolTypeEncodable<T::SolType>] alloc::sync::Arc<T> => T::SolType [];
+}
+
+macro_rules! tuple_impls {
+    ($count:literal $($ty:ident),+) => {
+        impl<$($ty: Encodable,)+> Encodable for ($($ty,)+) {
+            type SolType = ($($ty::SolType,)+);
+        }
+    };
+}
+
+impl Encodable for () {
+    type SolType = ();
+}
+
+all_the_tuples!(tuple_impls);
+
+#[cfg(test)]
+#[allow(clippy::type_complexity)]
+mod tests {
+    use super::*;
+
+    // Make sure these are in scope
+    #[allow(unused_imports)]
+    use crate::{SolType as _, SolTypeEncodable as _};
+
+    #[test]
+    fn inference() {
+        false.sol_type_name();
+        // false.abi_encoded_size();
+        // false.eip712_data_word();
+        // false.abi_encode_packed_to(&mut vec![]);
```

Comment:
These commented out methods are both in `SolTypeEncodable` and `Encodable`. Do we care enough to rename one of them?

Reply:
given that users are discouraged from importing `SolTypeEncodable` , i would skip the rename

Reply:
i guess this might make some difficulty for like, rust-analyzer auto-hints for imports? or for compiler messages?

Reply:
so i have revised my opinion in the last 2 minutes let's rename the internal ones, and keep the clear names for the user-facing one